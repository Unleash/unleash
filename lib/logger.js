'use strict';

const log4js = require('log4js');
log4js.configure({
    appenders: [{ type: 'console' }],
});

const logger = log4js.getLogger('unleash');

// TODO: make level configurable
if (process.env.NODE_ENV === 'production') {
    logger.setLevel(log4js.levels.ERROR);
} else if (process.env.NODE_ENV === 'test') {
    logger.setLevel(log4js.levels.ERROR);
} else {
    logger.setLevel(log4js.levels.DEBUG);
}

module.exports = logger;
