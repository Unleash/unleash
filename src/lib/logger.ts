import { configure, getLogger } from 'log4js';

export type LogProvider = (category?: string) => Logger;

export enum LogLevel {
    debug = 'debug',
    info = 'info',
    warn = 'warn',
    error = 'error',
    fatal = 'fatal',
}

export interface Logger {
    debug(message: any, ...args: any[]): void;
    info(message: any, ...args: any[]): void;
    warn(message: any, ...args: any[]): void;
    error(message: any, ...args: any[]): void;
    fatal(message: any, ...args: any[]): void;
}

export function getDefaultLogProvider(
    logLevel: LogLevel = LogLevel.error,
): LogProvider {
    configure({
        appenders: {
            console: { type: 'console' },
        },
        categories: {
            default: { appenders: ['console'], level: logLevel },
        },
    });

    return getLogger;
}

function validate(isValid: boolean, msg: string) {
    if (!isValid) {
        throw new TypeError(msg);
    }
}

export function validateLogProvider(provider: LogProvider): void {
    validate(typeof provider === 'function', 'Provider needs to be a function');

    const logger = provider('unleash:logger');
    validate(typeof logger.debug === 'function', 'Logger must implement debug');
    validate(typeof logger.info === 'function', 'Logger must implement info');
    validate(typeof logger.warn === 'function', 'Logger must implement warn');
    validate(typeof logger.error === 'function', 'Logger must implement error');
}
