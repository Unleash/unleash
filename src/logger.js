'use strict';

const log4js = require('log4js');

function getDefaultLogProvider() {
    let level;
    if (process.env.NODE_ENV === 'production') {
        level = log4js.levels.ERROR.levelStr;
    } else if (process.env.NODE_ENV === 'test') {
        level = log4js.levels.FATAL.levelStr;
    } else {
        level = log4js.levels.DEBUG.levelStr;
    }

    log4js.configure({
        appenders: {
            console: { type: 'console' },
        },
        categories: {
            default: { appenders: ['console'], level },
        },
    });

    return log4js.getLogger;
}

let loggerProvider = getDefaultLogProvider();

function validate(isValid, msg) {
    if (!isValid) {
        throw new TypeError(msg);
    }
}

module.exports.defaultLogProvider = loggerProvider;

function validateLogProvider(provider) {
    validate(typeof provider === 'function', 'Provider needs to be a function');

    const logger = provider('unleash:logger');
    validate(typeof logger.debug === 'function', 'Logger must implement debug');
    validate(typeof logger.info === 'function', 'Logger must implement info');
    validate(typeof logger.warn === 'function', 'Logger must implement warn');
    validate(typeof logger.error === 'function', 'Logger must implement error');
}

exports.validateLogProvider = validateLogProvider;

// Deprecated
exports.setLoggerProvider = function setLoggerProvider(provider) {
    validateLogProvider(provider);

    loggerProvider = provider;
    const logger = provider('unleash:logger');
    logger.info(`Your way of configuring a logProvider is depreacted. 
        See https://unleash.github.io/docs/getting_started for details`);
};
