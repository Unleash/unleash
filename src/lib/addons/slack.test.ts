import {
    FEATURE_CREATED,
    FEATURE_ARCHIVED,
    FEATURE_ENVIRONMENT_DISABLED,
    IEvent,
} from '../types/events';
import { Logger } from '../logger';

import SlackAddon from './slack';

import noLogger from '../../test/fixtures/no-logger';

let fetchRetryCalls: any[] = [];

jest.mock(
    './addon',
    () =>
        class Addon {
            logger: Logger;

            constructor(definition, { getLogger }) {
                this.logger = getLogger('addon/test');
                fetchRetryCalls = [];
            }

            async fetchRetry(url, options, retries, backoff) {
                fetchRetryCalls.push({
                    url,
                    options,
                    retries,
                    backoff,
                });
                return Promise.resolve({ status: 200 });
            }
        },
);

test('Should call slack webhook', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 1,
        createdAt: new Date(),
        type: FEATURE_CREATED,
        createdBy: 'some@user.com',
        project: 'default',
        featureName: 'some-toggle',
        data: {
            name: 'some-toggle',
            enabled: false,
            type: 'release',
            strategies: [{ name: 'default' }],
        },
    };

    const parameters = {
        url: 'http://hooks.slack.com',
    };

    await addon.handleEvent(event, parameters);
    expect(fetchRetryCalls.length).toBe(1);
    expect(fetchRetryCalls[0].url).toBe(parameters.url);
    expect(fetchRetryCalls[0].options.body).toMatchSnapshot();
});

test('Should call slack webhook for archived toggle', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 2,
        createdAt: new Date(),
        type: FEATURE_ARCHIVED,
        featureName: 'some-toggle',
        createdBy: 'some@user.com',
        data: {
            name: 'some-toggle',
        },
    };

    const parameters = {
        url: 'http://hooks.slack.com',
    };

    await addon.handleEvent(event, parameters);
    expect(fetchRetryCalls.length).toBe(1);
    expect(fetchRetryCalls[0].url).toBe(parameters.url);
    expect(fetchRetryCalls[0].options.body).toMatchSnapshot();
});

test('Should call slack webhook for archived toggle with project info', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 2,
        createdAt: new Date(),
        type: FEATURE_ARCHIVED,
        featureName: 'some-toggle',
        project: 'some-project',
        createdBy: 'some@user.com',
        data: {
            name: 'some-toggle',
        },
    };

    const parameters = {
        url: 'http://hooks.slack.com',
    };

    await addon.handleEvent(event, parameters);
    expect(fetchRetryCalls.length).toBe(1);
    expect(fetchRetryCalls[0].url).toBe(parameters.url);
    expect(fetchRetryCalls[0].options.body).toMatchSnapshot();
});

test(`Should call webhook for toggled environment`, async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 2,
        createdAt: new Date(),
        type: FEATURE_ENVIRONMENT_DISABLED,
        createdBy: 'some@user.com',
        environment: 'development',
        project: 'default',
        featureName: 'some-toggle',
        data: {
            name: 'some-toggle',
        },
    };

    const parameters = {
        url: 'http://hooks.slack.com',
    };

    await addon.handleEvent(event, parameters);
    expect(fetchRetryCalls).toHaveLength(1);
    expect(fetchRetryCalls[0].url).toBe(parameters.url);
    expect(fetchRetryCalls[0].options.body).toMatch(/disabled/);
    expect(fetchRetryCalls[0].options.body).toMatchSnapshot();
});

test('Should use default channel', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 3,
        createdAt: new Date(),
        type: FEATURE_CREATED,
        createdBy: 'some@user.com',
        featureName: 'some-toggle',
        data: {
            name: 'some-toggle',
            enabled: false,
            strategies: [{ name: 'default' }],
        },
    };

    const parameters = {
        url: 'http://hooks.slack.com',
        defaultChannel: 'some-channel',
    };

    await addon.handleEvent(event, parameters);

    const req = JSON.parse(fetchRetryCalls[0].options.body);

    expect(req.channel).toBe('#some-channel');
});

test('Should override default channel with data from tag', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 4,
        createdAt: new Date(),
        type: FEATURE_CREATED,
        createdBy: 'some@user.com',
        featureName: 'some-toggle',
        data: {
            name: 'some-toggle',
            enabled: false,
            strategies: [{ name: 'default' }],
        },
        tags: [
            {
                type: 'slack',
                value: 'another-channel',
            },
        ],
    };

    const parameters = {
        url: 'http://hooks.slack.com',
        defaultChannel: 'some-channel',
    };

    await addon.handleEvent(event, parameters);

    const req = JSON.parse(fetchRetryCalls[0].options.body);

    expect(req.channel).toBe('#another-channel');
});

test('Should post to all channels in tags', async () => {
    const addon = new SlackAddon({
        getLogger: noLogger,
        unleashUrl: 'http://some-url.com',
    });
    const event: IEvent = {
        id: 5,
        createdAt: new Date(),
        type: FEATURE_CREATED,
        createdBy: 'some@user.com',
        featureName: 'some-toggle',
        data: {
            name: 'some-toggle',
            enabled: false,
            strategies: [{ name: 'default' }],
        },
        tags: [
            {
                type: 'slack',
                value: 'another-channel-1',
            },
            {
                type: 'slack',
                value: 'another-channel-2',
            },
        ],
    };

    const parameters = {
        url: 'http://hooks.slack.com',
        defaultChannel: 'some-channel',
    };

    await addon.handleEvent(event, parameters);

    const req1 = JSON.parse(fetchRetryCalls[0].options.body);
    const req2 = JSON.parse(fetchRetryCalls[1].options.body);

    expect(fetchRetryCalls).toHaveLength(2);
    expect(req1.channel).toBe('#another-channel-1');
    expect(req2.channel).toBe('#another-channel-2');
});
