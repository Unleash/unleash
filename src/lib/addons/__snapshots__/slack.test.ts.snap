// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should call slack webhook 1`] = `"{"username":"Unleash","icon_emoji":":unleash:","text":"some@user.com created feature toggle <http://some-url.com/projects/default/features/some-toggle|some-toggle> in project *default*","channel":"#undefined","attachments":[{"actions":[{"name":"featureToggle","text":"Open in Unleash","type":"button","value":"featureToggle","style":"primary","url":"http://some-url.com/projects/default/features/some-toggle"}]}]}"`;

exports[`Should call slack webhook for archived toggle 1`] = `"{"username":"Unleash","icon_emoji":":unleash:","text":" some@user.com just archived feature toggle *<http://some-url.com/archive|some-toggle>*","channel":"#undefined","attachments":[{"actions":[{"name":"featureToggle","text":"Open in Unleash","type":"button","value":"featureToggle","style":"primary","url":"http://some-url.com/archive"}]}]}"`;

exports[`Should call slack webhook for archived toggle with project info 1`] = `"{"username":"Unleash","icon_emoji":":unleash:","text":" some@user.com just archived feature toggle *<http://some-url.com/projects/some-project/archive|some-toggle>*","channel":"#undefined","attachments":[{"actions":[{"name":"featureToggle","text":"Open in Unleash","type":"button","value":"featureToggle","style":"primary","url":"http://some-url.com/projects/some-project/archive"}]}]}"`;

exports[`Should call webhook for toggled environment 1`] = `"{"username":"Unleash","icon_emoji":":unleash:","text":"some@user.com *disabled* <http://some-url.com/projects/default/features/some-toggle|some-toggle> in *development* environment in project *default*","channel":"#undefined","attachments":[{"actions":[{"name":"featureToggle","text":"Open in Unleash","type":"button","value":"featureToggle","style":"primary","url":"http://some-url.com/projects/default/features/some-toggle"}]}]}"`;
